File: Dart/test/add_two_numbers_test.dart
import 'package:test/test.dart';
import 'package:leetcode/add_two_numbers.dart';

typedef TestList = List<({ListNode l1, ListNode l2, ListNode? want})>;

bool areEqual(ListNode? l1, ListNode? l2) {
  while (l1 != null && l2 != null) {
    if (l1.val != l2.val) {
      return false;
    }
    l1 = l1.next;
    l2 = l2.next;
  }
  return true;
}

void main() {
  TestList cases = [
    (
      l1: ListNode(2, ListNode(4, ListNode(3, null))),
      l2: ListNode(5, ListNode(6, ListNode(4, null))),
      want: ListNode(7, ListNode(0, ListNode(8, null)))
    ),
    (
      l1: ListNode(0, null),
      l2: ListNode(0, null),
      want: ListNode(0, null),
    ),
    (
      l1: ListNode(
        9,
        ListNode(
          9,
          ListNode(
            9,
            ListNode(
              9,
              ListNode(9, ListNode(9, ListNode(9, null))),
            ),
          ),
        ),
      ),
      l2: ListNode(
        9,
        ListNode(9, ListNode(9, ListNode(9, null))),
      ),
      want: ListNode(
        8,
        ListNode(
          9,
          ListNode(
            9,
            ListNode(
              9,
              ListNode(
                0,
                ListNode(0, ListNode(0, ListNode(1, null))),
              ),
            ),
          ),
        ),
      ),
    ),
  ];

  for (var tt in cases) {
    test('add two numbers', () {
      ListNode? result = Solution().addTwoNumbers(tt.l1, tt.l2);
      expect(areEqual(result, tt.want), true);
    });
  }
}

File: TypeScript/tests/add_two_numbers.test.ts
import { createList } from "../lib.ts";
import { addTwoNumbers } from "../src/add_two_numbers.ts";
import { assertEquals } from "@std/assert";

const cases = [
  {
    l1: createList([2, 4, 3]),
    l2: createList([5, 6, 4]),
    want: createList([7, 0, 8]),
  },
  {
    l1: createList([0]),
    l2: createList([0]),
    want: createList([0]),
  },
  {
    l1: createList([9, 9, 9, 9, 9, 9, 9]),
    l2: createList([9, 9, 9, 9]),
    want: createList([8, 9, 9, 9, 0, 0, 0, 1]),
  },
];

for (let i = 0; i < cases.length; i++) {
  const t = cases[i];

  Deno.test(`addTwoNumbers - test case ${i + 1}`, () => {
    const result = addTwoNumbers(t.l1, t.l2);
    assertEquals(result, t.want);
  });
}
